niceman/distributions/debian.py

```diff
@@ -368,93 +365,212 @@ def _find_all_sources(self):
+    def _get_pkgs_arch_and_version(self, pkg_dicts):
+        # Convert package names to name:arch format
+        queries = []
+        for p in pkg_dicts:
+            # Use "dpkg -s pkg" to get the installed version and arch
+            architecture = p.get("architecture")
+            name = p.get("name")
+            queries.append(name if not architecture
+                           else "%s:%s" % (name, architecture))
+
+        # Find out how many packages we can query at once
+        max_len = max([len(q) for q in queries])
+        num_queries = max((_MAX_LEN_CMDLINE - 13) // (max_len + 1), 1)
+
+        cmd_results = []
+        # Execute dpkg -s, parse, and collate the output
+        for subqueries in (queries[pos:pos + num_queries]
+                           for pos in range(0, len(queries), num_queries)):
+            try:
+                out, _ = self._session.execute_command(
+                    ['dpkg', '-s'] + subqueries
+                )
+                out = utils.to_unicode(out, "utf-8")
+                # dpkg -s uses the same output as apt-cache show pkg
+                cmd_results += parse_apt_cache_show_pkgs_output(out)
+            except CommandError as _:
+                raise  # Currently raise any exception we encounter
+
+        # Turn dpkg -s results into a lookup table by package name
+        cmd_results = self.create_lookup_from_apt_cache_show(cmd_results)
+
+        for p in pkg_dicts:
+            # find the version table in the parsed command results for the
+            # current package
+            r = cmd_results.get("%s:%s" % (p.get("name"),
+                                             p.get("architecture")))
+            if not r:
+                r = cmd_results.get(p.get("name"))
+            if not r:
+                lgr.warning("Was unable to run dpkg -s for %s" %
+                            p.get("name"))
+                continue
+            p["architecture"] = r.get("Architecture")
+            p["version"] = r.get("Version")
+
+    @staticmethod
+    def create_lookup_from_apt_cache_show(cmd_results):
+        lookup_results = {}
+        for r in cmd_results:
+            lookup_results[r.get("Package")] = r
+            lookup_results["%s:%s" % (r.get("Package"),
+                                      r.get("architecture"))] = r
+        return lookup_results
+
+    def _get_pkgs_details_from_apt_cache_show(self, pkg_dicts):
+        # Convert package names to name:arch format
+        queries = []
+        for p in pkg_dicts:
+            # Use "dpkg -s pkg" to get the installed version and arch
+            architecture = p.get("architecture")
+            name = p.get("name")
+            version = p.get("version")
+            queries.append("%s=%s" % (name, version) if not architecture
+                           else "%s:%s=%s" % (name, architecture, version))
+
+        # Find out how many packages we can query at once
+        max_len = max([len(q) for q in queries])
```

Comment:
minor (feel free to ignore, just nitpicking): `max(map(len, queries))`  or use `()` comprehension not a `[]` to avoid creating a possibly long list (not that it would make it much faster, but who knows??)

Reply:
Done (now in utils.py get_cmd_batch_len() )